src/injected/cssSelectorEngine.ts

```diff
@@ -87,73 +87,99 @@ export function createCSSEngine(shadow: boolean): SelectorEngine {
       //   return simple;
       // if (!shadow)
       //   return;
-      const parts = split(selector);
-      if (!parts.length)
+      const selectors = split(selector).filter(parts => !!parts.length).map(parts => parts.reverse());
+      if (!selectors.length)
         return;
-      parts.reverse();
-      return queryShadowInternal(root, root, parts, shadow);
+      return queryShadowInternal(root, root, selectors, shadow);
     },
 
     queryAll(root: SelectorRoot, selector: string): Element[] {
       // TODO: uncomment for performance.
       // if (!shadow)
       //   return Array.from(root.querySelectorAll(selector));
       const result: Element[] = [];
-      const parts = split(selector);
-      if (parts.length) {
-        parts.reverse();
-        queryShadowAllInternal(root, root, parts, shadow, result);
-      }
+      const selectors = split(selector).filter(parts => !!parts.length).map(parts => parts.reverse());
+      if (selectors.length)
+        queryShadowAllInternal(root, root, selectors, shadow, result);
       return result;
     }
   };
   (engine as any)._test = () => test(engine);
   return engine;
 }
 
-function queryShadowInternal(boundary: SelectorRoot, root: SelectorRoot, parts: string[], shadow: boolean): Element | undefined {
-  const matching = root.querySelectorAll(parts[0]);
-  for (let i = 0; i < matching.length; i++) {
-    const element = matching[i];
-    if (parts.length === 1 || matches(element, parts, boundary))
-      return element;
+function queryShadowInternal(boundary: SelectorRoot, root: SelectorRoot, selectors: string[][], shadow: boolean): Element | undefined {
+  let elements: NodeListOf<Element> | undefined;
+  if (selectors.length === 1) {
+    const parts = selectors[0];
+    const matching = root.querySelectorAll(parts[0]);
+    for (let i = 0; i < matching.length; i++) {
+      const element = matching[i];
+      if (parts.length === 1 || ancestorsMatch(element, parts, boundary))
```

Comment:
This is equivalent right? I'm having trouble tracking the special casings of 1 and 0
```suggestion
      if (ancestorsMatch(element, parts, boundary))
```

Reply:
Done.